freud/diffraction.pyx

```diff
@@ -23,11 +23,248 @@ import freud.locality
+    :math:`\text{sinc}(k r)`. For a full derivation see :cite:`Farrow2009`.
+    Note that the definition requires :math:`S(0) = N`.
+
+    The Debye implementation provides a much faster algorithm, but gives worse
+    results than :py:attr:`freud.diffraction.StaticStructureFactorDirect`
+    at low-k values.
+
+    .. note::
+        This code assumes all particles have a form factor :math:`f` of 1.
+
+    Partial structure factors can be computed by providing a set of
+    ``query_points`` and the total number of points in the system ``N_total`` to
+    the :py:meth:`compute` method. The normalization criterion is based on the
+    Faber-Ziman formalism. For particle types :math:`\alpha` and :math:`\beta`,
+    we compute the total scattering function as a sum of the partial scattering
+    functions as:
+
+    .. math::
+
+        S(k) - 1 = \sum_{\alpha}\sum_{\beta} \frac{N_{\alpha}
+        N_{\beta}}{N_{total}^2} \left(S_{\alpha \beta}(k) - 1\right)
+
+    Args:
+        bins (unsigned int):
+            Number of bins in :math:`k` space.
+        k_max (float):
+            Maximum :math:`k` value to include in the calculation.
+        k_min (float, optional):
+            Minimum :math:`k` value included in the calculation. Note that
+            there are practical restrictions on the validity of the
+            calculation in the long-wavelength regime,
+            see :py:attr:`min_valid_k`
+            (Default value = 0).
+    """
+    cdef freud._diffraction.StaticStructureFactorDebye * thisptr
+
+    def __cinit__(self, unsigned int bins, float k_max, float k_min=0):
+        if type(self) == StaticStructureFactorDebye:
+            self.thisptr = new freud._diffraction.StaticStructureFactorDebye(
+                bins, k_max, k_min)
+
+    def __dealloc__(self):
+        if type(self) == StaticStructureFactorDebye:
+            del self.thisptr
+
+    def compute(self, system, query_points=None, N_total=None, reset=True):
+        r"""Computes static structure factor.
+
+        Example for a single component system::
+
+            >>> sf = freud.diffraction.StaticStructureFactorDebye(
+            ...     bins=100, k_max=10, k_min=0
+            ... )
+            >>> sf.compute((box, points))
+
+        Example for partial mixed structure factor for a multiple component
+        system with types A and B::
+
+            >>> sf = freud.diffraction.StaticStructureFactorDebye(
+            ...     bins=100, k_max=10, k_min=0
+            ... )
+            >>> sf.compute(
+            ...     system=(box, A_points),
+            ...     query_points=B_points,
+            ...     N_total=N_particles
+            ... )
+
+        Args:
+            system:
+                Any object that is a valid argument to
+                :class:`freud.locality.NeighborQuery.from_system`.
+                Note that box is allowed to change when calculating trajectory
+                average static structure factor.
```

Comment:
I'm unsure about this comment:
> Note that box is allowed to change when calculating trajectory average static structure factor.

I'm leaving it in but I think it's a little awkward. We discuss accumulating and resetting, but usually not "trajectory average." The accumulation/reset doesn't necessarily have to be over an MD trajectory (could be independent replicas, or Monte Carlo which isn't really a trajectory, etc.).

Reply:
I agree that this doesn't seem to fit the general language of freud. I will change this in the new PR. 

I can't see why NPT simulations would produce incorrect results. Especially in this implementation where there is no underlying k-space grid but you histogram distances directly which (aside from applying PBCs correctly) shouldn't care about box properties. The box properties are supposed to fluctuate around some equilibrium value anyways if a proper ensamble is sampled so this shouldn't be a problem. Is there anything that bothers you in particular here? We could also test this - by calculating S(q) for a system at same thermodynamic state point using NPT and NVT and comparing. But if proper equilibrium is reached in both there is no way that results would differ. This would be equivalent to some other structural property being different between NPT and NVT simulation under same conditions which shouldn't be the case(?).

As for calculations without PBCs, these probably don't make much sense, aside from possibly having a HUGE experimental sample of couple of millions ++ of particles  where one would want to calculate scattering numerically from positions of particles.

Reply:
I guess I don’t have specific objections aside from inconsistent language. I know our accumulation code for RDF and several other classes has to normalize by the box volume, and that means that we require constant box volume for correct results (we could improve this behavior by tracking more info about each frame’s box and perhaps reducing data over threads for each frame?).

Reply:
I see. That makes sense in those cases. S(k) doesn't care about volume right, so it shouldn't be an issue here. 

As for other functions, perhaps a good approach would be to track if the box changes in the compute function and then reduce only when box change occurs? Would such an approach make more sense? Then box wouldn't have to be stored for each step or reduction happens only when it is actually needed?

Reply:
Yes, I think this might be fine since it doesn't normalize by box volumes. I think it is possible to make changing boxes work for RDF and other functions using a strategy like you describe, but I think it would require careful attention. See also #616 for previous conversations on this topic.